### PR TITLE
Hdf5

### PR DESCRIPTION
**Pull request by dguibert**
###### Things done
- [X] Tested using sandboxing (`nix-build --option build-use-chroot true` or [nix.useChroot](http://nixos.org/nixos/manual/options.html#opt-nix.useChroot) on NixOS)
- Built on platform(s)
  - [X] NixOS
  - [ ] OS X
  - [ ] Linux
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: pkgs/tools/misc/hdf5/default.nix [hunk 1]
```diff
@@ -15,34 +17,45 @@ assert !cpp || mpi == null;
...
+    "-DCMAKE_BUILD_TYPE=Release"
+    "-DCMAKE_C_FLAGS=-fPIC"
+    "-DHDF5_BUILD_HL_LIB=ON"
+    "-DHDF5_DISABLE_COMPILER_WARNINGS=ON"
+    "-DHDF5_BUILD_TOOLS=ON"
+    "-DCMAKE_EXE_LINKER_FLAGS=''"
+    "-DHDF5_INSTALL_DATA_DIR=share/hdf5"
+    "-DHDF5_INSTALL_CMAKE_DIR=share/cmake/hdf5"
+  ]
+    ++ stdenv.lib.optional (mpi != null)  "-DHDF5_ENABLE_PARALLEL=ON -DHDF5_BUILD_CPP_LIB:BOOL=OFF"
```
Comment (lancelotsix) [inline pkgs/tools/misc/hdf5/default.nix]: Could the `HDF5_BUILD_CPP_LIB` depend on the `cpp` parameter (see line 5)?

With probably something like `assert mpi -> !cpp`

**Comment by lancelotsix** [comment]
I am running `nox-review pr 14622` locally to check if it works properly.

I also have external projects that to build against that version with CMake:

```bash
$ NIX_PATH=nixpkgs=/path/to/nixpkgs nix-build
[…]
CMake Error at /nix/store/7jsaqfrwhfrmfv2x3sjzx4rvyqgv2wmv-cmake-3.4.3/share/cmake-3.4/Modules/FindHDF5.cmake:234 (get_target_property):
  get_target_property() called with non-existent target "hdf5_cpp".
Call Stack (most recent call first):
  CMakeLists.txt:18 (find_package)


CMake Error at /nix/store/7jsaqfrwhfrmfv2x3sjzx4rvyqgv2wmv-cmake-3.4.3/share/cmake-3.4/Modules/FindPackageHandleStandardArgs.cmake:148 (message):
  Could NOT find HDF5 (missing: HDF5_LIBRARIES) (found version "1.10.0")
Call Stack (most recent call first):
  /nix/store/7jsaqfrwhfrmfv2x3sjzx4rvyqgv2wmv-cmake-3.4.3/share/cmake-3.4/Modules/FindPackageHandleStandardArgs.cmake:388 (_FPHSA_FAILURE_MESSAGE)
  /nix/store/7jsaqfrwhfrmfv2x3sjzx4rvyqgv2wmv-cmake-3.4.3/share/cmake-3.4/Modules/FindHDF5.cmake:398 (find_package_handle_standard_args)
  CMakeLists.txt:18 (find_package)
```

I’ll try to figure out what is happening ASAP (this is not a bloker, just an informal testcase for me)

**Comment by dguibert** [comment]
for some reason, cmake is finding hdf5 by its module "FindHDF5.cmae" and not "hdf5-config.cmake" provided by the package hdf5-cpp.

With this CMakeLists.txt, cmake does not complain:
```cmake
cmake_minimum_required(VERSION 3.4)
project(HDF5CPP)
find_package(HDF5 COMPONENTS CXX NO_MODULE)
```

**Comment by pipping** [comment]
I think the missing header issue is fixed in hdf5 1.10.0-patch1.

While upstream appears to allow you to choose between the autotools and cmake, I would highly recommend against using cmake here. the cmake option is first and foremost there to add support for the windows platform and on quite a few occasions I've walked away with the impression linux was far more thoroughly tested with the autotools-based system. Binaries or headers would be missing with cmake on linux. I've reported the issues I noticed and they were addressed but I'm not sure there I didn't miss anything.

Why do I even care? Isn't the outcome all the same? That's how it is for LLVM e.g., right? Well, in this case, no, not at all. If you use cmake to build, your shared library will end up with such a name that you cannot link to it via -lhdf. You would instead have to use -lhdf5-static or -lhdf5-shared. The list of headers that are installed is very different, the defines that make it into H5pubconf.h are very different (I can provide more details on this if you're interested). The list goes on.

**Comment by pipping** [comment]
@dguibert This slightly cryptic matter is down to the use of `NAMES`. The README on how to use HDF5 with cmake says that you should use the cmake check
```
find_package(HDF5 NAMES hdf5 COMPONENTS C static)
```
The effect the keyword `NAMES` in conjunction with `hdf5` here means that `FindHDF5.cmake` (installed by cmake) is ignored and `hdf5-config.cmake` is used instead. The latter is installed by hdf5 if you use the cmake build system.

So you have the choice between using
```
find_package(HDF5 NAMES hdf5)
```
and
```
find_package(HDF5)
```
The former will always work and just use the cmake-provided configuration. The latter will only work if hdf5 was built via cmake. Does the latter provide any advantage over the former? Not that I'm aware of. So why does `hdf5-config.cmake` even exist? I don't know.

**Comment by joachifm** [comment]
What is the status of this?

**Comment by dguibert** [comment]
I've updated the version of hdf5 and revert the use of cmake.

**Comment by FRidh** [comment]
Pushed 1c1a03f0d38606fd4a80e61a19ddfd3c463d99df and dfb3707dab86c852991d0668db2f46f0e3ba0ed9

**Pull request closed without merging**